### PR TITLE
 Add shortcuts support and automatic moving to next code selector

### PR DESCRIPTION
**Pull request by marianamarasoiu**
Sorry for the big PR, please have a look and let me know what you think of the proposed logic around processing user input (rows 60-137 in main_ui and rows 43 to 67 in view_model). The rest is primarily variable renames. Thanks!

### CODE REVIEW

File: webapp/lib/view_model.dart [hunk 1]
```diff
@@ -39,63 +34,86 @@ class MessageViewModel {
...
+  static CodeSelector get activeCodeSelector => _activeCodeSelector;
+  static set activeCodeSelector(CodeSelector activeCodeSelector) {
+    _activeCodeSelector?.viewElement?.classes?.toggle('active', false);
+    // Focus on the new code selector
+    _activeCodeSelector = activeCodeSelector;
+    _activeCodeSelector.viewElement.classes.toggle('active', true);
+    _activeCodeSelector.dropdown.focus();
+  }
+
+  static final String emptyCodeValue = 'unassign';
```
Comment (lukechurch) [inline webapp/lib/view_model.dart]: const EMPTY_CODE_VALUE

File: webapp/lib/main_ui.dart [hunk 2]
```diff
@@ -74,3 +161,15 @@ class CodaUI {
     this.tableBody?.remove();
   }
 }
+
+parentOfClass(Element element, String parentClass) {
```
Comment (lukechurch) [inline webapp/lib/main_ui.dart]: getElementWithParentOfClass

Comment (lukechurch) [reply webapp/lib/main_ui.dart]: Please could you also add a return type

File: webapp/lib/main_ui.dart [hunk 3]
```diff
@@ -45,9 +53,88 @@ class CodaUI {
...
       MessageViewModel messageViewModel = new MessageViewModel(message, dataset);
       messages.add(messageViewModel);
+      messageMap[message.id] = messageViewModel;
       body.append(messageViewModel.viewElement);
     });
     messageCodingTable.append(body);
+
+    messageCodingTable.onChange.listen((event) {
+      var target = event.target;
+      if (target is InputElement || target is SelectElement) {
```
Comment (lukechurch) [inline webapp/lib/main_ui.dart]: Can we change this to: ``` if (target is! TargetElement and target is! SelectElement) return; ```

**Comment by marianamarasoiu** [comment]
Here's the demo: https://www.cl.cam.ac.uk/~mcm79/codav2_demos/tabbing-2018-10-07/

File: webapp/lib/main_ui.dart (hunk 2 above)
Comment (marianamarasoiu) [reply webapp/lib/main_ui.dart]: Done, thanks!

File: webapp/lib/view_model.dart (hunk 1 above)
Comment (marianamarasoiu) [reply webapp/lib/view_model.dart]: Done, thanks!

File: webapp/lib/main_ui.dart (hunk 3 above)
Comment (marianamarasoiu) [reply webapp/lib/main_ui.dart]: Done, thanks!

**Comment by marianamarasoiu** [comment]
Thanks for this! I've resolved all your comments so far, please have another look!

File: webapp/lib/main_ui.dart [hunk 4]
```diff
@@ -45,9 +53,88 @@ class CodaUI {
...
+      if (target is! InputElement && target is! SelectElement) return;
+
+      TableRowElement row = getAncestors(CodeSelector.activeCodeSelector.viewElement).firstWhere((e) => e.classes.contains('message-row'));
+      DivElement inputGroup = getAncestors(CodeSelector.activeCodeSelector.viewElement).firstWhere((e) => e.classes.contains('input-group'));
+      String messageID = row.attributes['message-id'];
+      MessageViewModel message = messageMap[messageID];
+      String schemeID = inputGroup.attributes['scheme-id'];
+
+      if (target is InputElement) { // change on checkbox element
+        message.schemeCheckChanged(dataset, schemeID, target.checked);
```
Comment (lukechurch) [inline webapp/lib/main_ui.dart]: I think you can add a shortcut return here, or turn the if statements into a switch

File: webapp/lib/main_ui.dart [hunk 5]
```diff
@@ -45,9 +53,88 @@ class CodaUI {
...
+        CodeSelector codeSelector = message.codeSelectors.singleWhere((codeSelector) => codeSelector.scheme.id == schemeID);
+        CodeSelector.activeCodeSelector = codeSelector;
+        message.schemeCodeChanged(dataset, schemeID, codeSelector.selectedOption);
+        codeSelector.hideWarning();
+        selectNextEmptyCodeSelector(messageID, schemeID);
+      }
+    });
+
+    window.onKeyDown.listen((event) {
+      if (event.key == 'Tab') {
```
Comment (lukechurch) [inline webapp/lib/main_ui.dart]: if != tab, return.

File: webapp/lib/main_ui.dart [hunk 6]
```diff
@@ -45,9 +53,88 @@ class CodaUI {
...
+        selectNextEmptyCodeSelector(messageId, CodeSelector.activeCodeSelector.scheme.id);
+        event.preventDefault();
+        event.stopPropagation();
+        return;
+      }
+    });
+    CodeSelector.activeCodeSelector = messages[0].codeSelectors[0];
+  }
+
+  selectNextEmptyCodeSelector(String messageID, String schemeID) {
```
Comment (lukechurch) [inline webapp/lib/main_ui.dart]: I think splitting this into 2 methods, selectNextEmptyCodeSelectorVertical and selectNextEmptyCodeSelectorHorizontal would make sense and help decrease the level of if-else nesting

File: webapp/lib/main_ui.dart [hunk 7]
```diff
@@ -74,3 +161,12 @@ class CodaUI {
     this.tableBody?.remove();
   }
 }
+
+List<Element> getAncestors(Element element) {
```
Comment (lukechurch) [inline webapp/lib/main_ui.dart]: Much easier to understand, thanks

File: webapp/lib/main_ui.dart (hunk 6 above)
Comment (marianamarasoiu) [reply webapp/lib/main_ui.dart]: That's a good idea, thanks!

File: webapp/lib/main_ui.dart (hunk 5 above)
Comment (marianamarasoiu) [reply webapp/lib/main_ui.dart]: The if statement isn't for the whole function, only for the tab key. I'm handling other keys below. Am I missing something?

File: webapp/lib/main_ui.dart (hunk 4 above)
Comment (marianamarasoiu) [reply webapp/lib/main_ui.dart]: I'll add a shortcut return, switching would result in losing the type of the target element useful for not needing to do a cast for target.checked.

File: webapp/lib/main_ui.dart (hunk 5 above)
Comment (lukechurch) [reply webapp/lib/main_ui.dart]: Ah, I see, I got confused by the closure sorry

**Pull request merged**